chrome/content/zotero/xpcom/feedReader.js

```diff
@@ -0,0 +1,538 @@
+ *   representing feed properties
+ * @property {Zotero.Promise<FeedItem>*} ItemIterator Returns an iterator
+ *   for feed items. The iterator returns FeedItem promises that have to be
+ *   resolved before requesting the next promise. When all items are exhausted.
+ *   the promise resolves to null.
+ * @method {void} terminate Stops retrieving/parsing the feed. Data parsed up
+ *   to this point is still available.
+ */
+Zotero.FeedReader = function(url) {
+	if (!url) throw new Error("Feed URL must be supplied");
+
+	
+	this._url = url;
+	this._feedItems = [Zotero.Promise.defer()];
+	this._feedProcessed = Zotero.Promise.defer();
+
+	let feedFetched = Zotero.Promise.defer();
+	feedFetched.promise.then(function(feed) {
+		let info = {};
+		
+		info.title = feed.title ? feed.title.plainText() : '';
+		info.subtitle = feed.subtitle ? feed.subtitle.plainText() : '';
+		
+		if (feed.updated) info.updated = new Date(feed.updated);
+		
+		// categories: MDN says "not yet implemented"
+		
+		info.creators = Zotero.FeedReader._processCreators(feed, 'authors', 'author');
+		
+		// TODO: image as icon
+		
+		let publicationTitle = Zotero.FeedReader._getFeedField(feed, 'publicationName', 'prism')
+			|| Zotero.FeedReader._getFeedField(feed, 'pubTitle');
+		if (publicationTitle) info.publicationTitle = publicationTitle;
+		
+		let publisher = Zotero.FeedReader._getFeedField(feed, 'publisher', 'dc');
+		if (publisher) info.publisher = publisher;
+		
+		let rights = (feed.rights && feed.rights.plainText())
+			|| Zotero.FeedReader._getFeedField(feed, 'copyright', 'prism')
+			|| Zotero.FeedReader._getFeedField(feed, 'rights', 'dc')
+			|| Zotero.FeedReader._getFeedField(feed, 'copyright');
+		if (rights) info.rights = rights;
+		
+		let issn = Zotero.FeedReader._getFeedField(feed, 'issn', 'prism');
+		if (issn) info.ISSN = issn;
+		
+		let isbn = Zotero.FeedReader._getFeedField(feed, 'isbn', 'prism')
+			|| Zotero.FeedReader._getFeedField(feed, 'isbn')
+		if (isbn) info.ISBN = isbn;
+		
+		let language = Zotero.FeedReader._getFeedField(feed, 'language', 'dc')
+			|| Zotero.FeedReader._getFeedField(feed, 'language');
+		if (language) info.language = language;
+		
+		let ttl = Zotero.FeedReader._getFeedField(feed, 'ttl');
+		if (ttl) info.ttl = ttl;
+		
+		this._feedProperties = info;
+		this._feed = feed;
+		return info;
+	}.bind(this)).then(function(){
+		let items = this._feed.items;
+		if (items && items.length) {
+			for (let i=0; i<items.length; i++) {
+				let item = items.queryElementAt(i, Components.interfaces.nsIFeedEntry);
+				if (!item) continue;
+				
+				let feedItem = Zotero.FeedReader._getFeedItem(item, this._feedProperties);
+				if (!feedItem) continue;
+				
+				let lastItem = this._feedItems[this._feedItems.length - 1];
+				this._feedItems.push(Zotero.Promise.defer()); // Push a new deferred promise so an iterator has something to return
+				lastItem.resolve(feedItem);
+			}
+
+		this._feedProcessed.resolve();
+		}
```

Comment:
Something doesn't look right here. I think `this._feedProcessed.resolve();` and `}` are flipped here (that would account for the indenting issue too). Also, we need a `.catch` before the `.finally` that would `.reject()` the `this._feedProcessed`, otherwise things might hang if there's an error.